content/assignments/mini-libc/tests/test_malloc.sh

```diff
@@ -0,0 +1,56 @@
+#!/bin/bash
+
+# SPDX-License-Identifier: BSD-3-Clause
+
+source graded_test.inc.sh
+
+exec_file=./memory/malloc
+
+test_malloc()
+{
+    if test ! -f "$exec_file"; then
+        echo "No such file $exec_file" 1>&2
+        exit 1
+    fi
+
+    nm "$exec_file" > /dev/null 2>&1
+    if test $? -ne 0; then
+        echo "No symbols in file" 1>&2
+        exit 1
+    fi
+
+    nm "$exec_file" | grep ' malloc$' > /dev/null
+    if test $? -ne 0; then
+        echo "No malloc symbol" 1>&2
+        exit 1
+    fi
+
+    nm "$exec_file" | grep ' mmap$' > /dev/null
+    if test $? -ne 0; then
+        echo "No mmap symbol" 1>&2
+        exit 1
+    fi
+
+    timeout -k 1 2 strace "$exec_file" 2>&1 > /dev/null | grep mmap > /dev/null
+    if test $? -ne 0; then
+        echo "No mmap syscall" 1>&2
+        exit 1
+    fi
+
+    "$exec_file" &
+    PID=$!
```

Comment:
Also, I noticed that without an extra sleep less than 1 second added here (to account for the 1s sleep in the source file), the behaviour is unexpected, generating different results on different runs, not sure if it is a problem only on my end but the extra sleep generates a stable response on multiple runs

Reply:
this was also what I noticed.

Reply:
@pufi656, @JunisvaultCo, I now understand the issue. Will be back with a fix, please then let me know if it works on your side.

Reply:
@razvand, can't it be also improved by using interruption signals?

Reply:
@pufi656, @JunisvaultCo, please see if the updated version works for you.

Reply:
I've investigated this issue by putting pmaps just before mem1 (and echoing mem1 and mem2) and it looks as if somehow sometimes it just takes the mem count of the bash, therefore messing up the first mem1 by making it big.
The files that use mem1 and mem2 in such a way are:

test_malloc_free_sequence.sh
test_malloc_free.sh
test_malloc.sh
test_mmap_munmap.sh
test_mmap.sh
test_multiple_malloc_free.sh
test_multiple_malloc.sh

Reply:
@pufi656, @JunisvaultCo, [this commit](https://github.com/open-education-hub/operating-systems/pull/209/commits/25901577fb8a2420209484343d277fb63fceaa93) fixes that. Let me know if it works on your side. I also added issue #227 to provide a future full-proof fix.

Could you also provide your full names and e-mail addresses so I can add you as co-authors?

Reply:
> @razvand, can't it be also improved by using interruption signals?

No. This is a minimal libc that doesn't implement signals; that would make it too complicated.

Reply:
> I've investigated this issue by putting pmaps just before mem1 (and echoing mem1 and mem2) and it looks as if somehow sometimes it just takes the mem count of the bash, therefore messing up the first mem1 by making it big. The files that use mem1 and mem2 in such a way are:
> 
> test_malloc_free_sequence.sh test_malloc_free.sh test_malloc.sh test_mmap_munmap.sh test_mmap.sh test_multiple_malloc_free.sh test_multiple_malloc.sh

I didn't understand the issue. Could you elaborate, ideally with a concrete example?

Reply:
test_malloc.sh is fine now. A similar sleep (maybe sleep 0.5 to not increase by that much) before PID=$! should be added in all the files I mentioned above.

Reply:
Adding 
    pmap $PID

before 
    mem1=$(ps -o vsz --noheader --pid $PID)
will usually give:

4425:   /bin/bash ./test_multiple_malloc.sh
00005569ed400000   1040K r-x-- bash
00005569ed703000     16K r---- bash
00005569ed707000     36K rw--- bash
00005569ed710000     40K rw---   [ anon ]
00005569ed766000    132K rw---   [ anon ]
00005569ed787000    132K rw---   [ anon ]
00007f87a6ea0000   1484K r---- LC_COLLATE
00007f87a7013000   1948K r-x-- libc-2.27.so
00007f87a71fa000   2048K ----- libc-2.27.so
00007f87a73fa000     16K r---- libc-2.27.so
00007f87a73fe000      8K rw--- libc-2.27.so
00007f87a7400000     16K rw---   [ anon ]
00007f87a7404000     12K r-x-- libdl-2.27.so
00007f87a7407000   2044K ----- libdl-2.27.so
00007f87a7606000      4K r---- libdl-2.27.so
00007f87a7607000      4K rw--- libdl-2.27.so
00007f87a7608000    148K r-x-- libtinfo.so.5.9
00007f87a762d000   2048K ----- libtinfo.so.5.9
00007f87a782d000     16K r---- libtinfo.so.5.9
00007f87a7831000      4K rw--- libtinfo.so.5.9
00007f87a7832000    164K r-x-- ld-2.27.so
00007f87a7876000    196K r---- LC_CTYPE
00007f87a78a7000   1644K r---- locale-archive
00007f87a7a42000     20K rw---   [ anon ]
00007f87a7a4a000      4K r---- LC_NUMERIC
00007f87a7a4b000      4K r---- LC_TIME
00007f87a7a4c000      4K r---- LC_MONETARY
00007f87a7a4d000      4K r---- SYS_LC_MESSAGES
00007f87a7a4e000      4K r---- LC_PAPER
00007f87a7a4f000      4K r---- LC_NAME
00007f87a7a50000      4K r---- LC_ADDRESS
00007f87a7a51000      4K r---- LC_TELEPHONE
00007f87a7a52000      4K r---- LC_MEASUREMENT
00007f87a7a53000     28K r--s- gconv-modules.cache
00007f87a7a5a000      4K r---- LC_IDENTIFICATION
00007f87a7a5b000      4K r---- ld-2.27.so
00007f87a7a5c000      4K rw--- ld-2.27.so
00007f87a7a5d000      4K rw---   [ anon ]
00007fff8b66c000    132K rw---   [ stack ]
00007fff8b6fb000     16K r----   [ anon ]
00007fff8b6ff000      4K r-x--   [ anon ]

further adding 
    echo $mem1 $mem2
after 

    mem2=$(ps -o vsz --noheader --pid $PID)

will sometimes give
13452 12464
test_multiple_malloc             ........................ failed ...  0

That 13452 seems to be the memory size of the bash process that calls the executable.

Reply:
@JunisvaultCo, how are you testing this, are you using a virtual machine?

Reply:
@JunisvaultCo, retry now. I should have fixed this in [this commit](https://github.com/open-education-hub/operating-systems/pull/209/commits/c8828d6f662f28082dbdb825c5b5466f83cabc51).

Reply:
@pufi656, @vladtf, also pinging you to check this on your side.

Reply:
@pufi656, what about the othe `malloc` / `mmap` scripts?

Reply:
@razvand, I've checked with latest changes from this pr an right now only these 2 tests are failing:
test_multiple_malloc_free        ........................ failed ...  0
test_malloc_free_sequence        ........................ failed ...  0

I will check what is the cause, but probably the same.

Reply:
Nvm, everything seems to be working fine, I should clean the remotes and check again.

Reply:
@razvand confirming all scripts regarding memory testing work fine